kpops/utils/yaml_loading.py

```diff
@@ -37,3 +37,45 @@ def substitute(input: str, substitution: Mapping[str, Any] | None = None) -> str
     if not substitution:
         return input
     return Template(input).safe_substitute(**substitution)
+
+
+def substitute_nested(input: str, **kwargs) -> str:
+    """Allow for multiple substitutions to be passed.
+
+    Will make as many passes as needed to substitute all possible placeholders.
+    A ceiling is set to avoid infinite loops.
+
+    HINT: If :param input: is a ``Mapping`` that you converted into ``str``,
+    You can pass it as a string, and as a ``Mapping`` to enable self-reference.
+
+    :Example:
+
+    >>> substitution = {
+            "a": "0",
+            "b": "${a}",
+            "c": "${b}",
+            "d": "${a}",
+        }
+    >>> input = "${a}, ${b}, ${c}, ${d}"
+    >>> print("Substituted string: " + substitute_nested(input, **substitution))
+    0, 0, 0, 0
+
+    :param input: The raw input containing $-placeholders
+    :type input: str
+    :param **kwargs: Substitutions
+    :raises Exception: An infinite loop condition detected. Check substitution variables.
+    :return: Substituted input string
+    :rtype: str
+    """
+    if not kwargs:
+        return input
+    old_str, new_str = "", substitute(input, kwargs)
+    steps = set()
+    while new_str not in steps:
+        steps.add(new_str)
+        old_str, new_str = new_str, substitute(new_str, kwargs)
+    if new_str != old_str:
+        raise Exception(
```

Comment:
we should avoid raising generic `Exception`, it's an anti-pattern. how about `ValueError`, i.e. input is invalid?